channeldb/graph_cache.go

```diff
@@ -0,0 +1,331 @@
+		policy2 != nil && policy2.IsDisabled() {
+
+		return
+	}
+
+	// Create the edge entry for both nodes.
+	c.mtx.Lock()
+	c.updateOrAddEdge(info.NodeKey1Bytes, &DirectedChannel{
+		ChannelID: info.ChannelID,
+		IsNode1:   true,
+		OtherNode: info.NodeKey2Bytes,
+		Capacity:  info.Capacity,
+	})
+	c.updateOrAddEdge(info.NodeKey2Bytes, &DirectedChannel{
+		ChannelID: info.ChannelID,
+		IsNode1:   false,
+		OtherNode: info.NodeKey1Bytes,
+		Capacity:  info.Capacity,
+	})
+	c.mtx.Unlock()
+
+	// The policy's node is always the to_node. So if policy 1 has to_node
+	// of node 2 then we have the policy 1 as seen from node 1.
+	if policy1 != nil {
+		fromNode, toNode := info.NodeKey1Bytes, info.NodeKey2Bytes
+		if policy1.Node.PubKeyBytes != info.NodeKey2Bytes {
+			fromNode, toNode = toNode, fromNode
+		}
+		isEdge1 := policy1.ChannelFlags&lnwire.ChanUpdateDirection == 0
+		c.UpdatePolicy(policy1, fromNode, toNode, isEdge1)
+	}
+	if policy2 != nil {
+		fromNode, toNode := info.NodeKey2Bytes, info.NodeKey1Bytes
+		if policy2.Node.PubKeyBytes != info.NodeKey1Bytes {
+			fromNode, toNode = toNode, fromNode
+		}
+		isEdge1 := policy2.ChannelFlags&lnwire.ChanUpdateDirection == 0
+		c.UpdatePolicy(policy2, fromNode, toNode, isEdge1)
+	}
+}
+
+// updateOrAddEdge makes sure the edge information for a node is either updated
+// if it already exists or is added to that node's list of channels.
+func (c *GraphCache) updateOrAddEdge(node route.Vertex, edge *DirectedChannel) {
+	for idx := range c.nodeChannels[node] {
+		// Do we already have the channel in question?
+		if c.nodeChannels[node][idx].ChannelID == edge.ChannelID {
+			c.nodeChannels[node][idx] = edge
+			return
+		}
+	}
+
+	// We didn't find the channel, let's add it.
+	c.nodeChannels[node] = append(c.nodeChannels[node], edge)
+}
+
+// UpdatePolicy updates a single policy on both the from and to node. The order
+// of the from and to node is not strictly important. But we assume that a
+// channel edge was added beforehand so that the directed channel struct already
+// exists in the cache.
+func (c *GraphCache) UpdatePolicy(policy *ChannelEdgePolicy, fromNode,
+	toNode route.Vertex, edge1 bool) {
+
+	// If a policy's node is nil, we can't cache it yet as that would lead
+	// to problems in pathfinding.
+	if policy.Node == nil {
+		// TODO(guggero): Fix this problem!
+		log.Warnf("Cannot cache policy because of missing node (from "+
+			"%x to %x)", fromNode[:], toNode[:])
+		return
+	}
+
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	updatePolicy := func(channels []*DirectedChannel) {
+		for idx, channel := range channels {
+			// Is this the channel in question?
+			if channel.ChannelID == policy.ChannelID {
+				// Edge 1 is defined as the policy for the
+				// direction of node1 to node2.
+				switch {
+				// This is node 1, and it is edge 1, so this is
+				// the outgoing policy for node 1.
+				case channel.IsNode1 && edge1:
+					channels[idx].OutPolicy = policy
```

Comment:
Yeah, you're right with all those nits. This code grew over many versions and I had some issues with pointers and function calls. But since we only do assignments here, we can skip the safety measure of using the index.